From a0ee38f7a4f4560fbb9b69b53ff02318eaeb645a Mon Sep 17 00:00:00 2001
From: Shubang Lu <shubang@google.com>
Date: Wed, 15 Apr 2026 12:12:08 -0700
Subject: [PATCH] [LiveTv] Avoid Intent Redirection to permission protected
 Activities

Test: mannual
Bug: 329233015
Flag: EXEMPT bugfix
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:415d636eb465a20ddcdd12244e675cdf18971f7b
Merged-In: I0f92a361c4f9b78c459905c06ecde5d9a2996c97
Change-Id: I0f92a361c4f9b78c459905c06ecde5d9a2996c97
---
 src/com/android/tv/data/ChannelImpl.java | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/src/com/android/tv/data/ChannelImpl.java b/src/com/android/tv/data/ChannelImpl.java
index 5be1179d..62c06cc2 100644
--- a/src/com/android/tv/data/ChannelImpl.java
+++ b/src/com/android/tv/data/ChannelImpl.java
@@ -681,11 +681,28 @@ public final class ChannelImpl implements Channel {
                     boolean isProtectedActivity = packageName != null
                             && (packageName.equals(CommonConstants.BASE_PACKAGE)
                             || packageName.startsWith(CommonConstants.BASE_PACKAGE + "."));
+
+                    // Prevent creation of App Links to permission protected activities
+                    boolean isPermissionProtected =
+                            activityInfo.exported && (activityInfo.permission != null);
+
                     if (isProtectedActivity) {
-                        Log.w(TAG,"Attempt to add app link to protected activity: "
+                        Log.w(TAG, "Attempt to add app link to protected activity: "
+                                + mAppLinkIntentUri);
+                        return;
+                    }
+                    if (isPermissionProtected){
+                        Log.w(TAG, "Attempt to add app link to permission protected activity: "
                                 + mAppLinkIntentUri);
                         return;
                     }
+                    // Note: Normally, we should consider to restrict ChooserIntent and ACTION_SEND
+                    // Intent as well, but
+                    // 1. ChooserIntent: with Intent.parseUri, an attacker cannot wrap
+                    // Intent.EXTRA_INTENT in the app link
+                    // 2. ACTION_SEND: After test, an attacker cannot use ACTION_SEND to get
+                    // FLAG_GRANT_* here
+
                     mAppLinkIntent = intent;
                     mAppLinkIntent.putExtra(
                             CommonConstants.EXTRA_APP_LINK_CHANNEL_URI, getUri().toString());
-- 
2.55.0.897.gb25b4bd76c-goog

